src/dotty/tools/dotc/util/DiffUtil.scala

```diff
@@ -30,38 +30,120 @@ object DiffUtil {
-          lines(pos) = old + ADDITION_COLOR + lines(pos)
-          lines(endPos) = lines(endPos) + ANSI_DEFAULT
+  private sealed trait Patch
+  private final case class Unmodified(str: String) extends Patch
+  private final case class Modified(original: String, str: String) extends Patch
+  private final case class Deleted(str: String) extends Patch
+  private final case class Inserted(str: String) extends Patch
 
-        case "DELETE" if printDiffDel =>
-          val deleted = delta.getOriginal.getLines.mkString
-          if (!deleted.forall(Character.isWhitespace)) {
-            lines(pos) = DELETION_COLOR + deleted + ANSI_DEFAULT + lines(pos)
-          }
+  private def hirschberg(a: Array[String], b: Array[String]): Array[Patch] = {
+    def build(x: Array[String], y: Array[String], builder: mutable.ArrayBuilder[Patch]): Unit = {
+      if (x.isEmpty) {
+        builder += Inserted(y.mkString)
+      } else if (y.isEmpty) {
+        builder += Deleted(x.mkString)
+      } else if (x.length == 1 || y.length == 1) {
+        needlemanWunsch(x, y, builder)
+      } else {
+        val xlen = x.length
+        val xmid = xlen / 2
+        val ylen = y.length
+
+        val (x1, x2) = x.splitAt(xmid)
+        val ScoreL = nwScore(x1, y)
+        val ScoreR = nwScore(x2.reverse, y.reverse)
+        val scoreSum = (ScoreL zip ScoreR.reverse).map(tup => tup._1 + tup._2)
+        val max = scoreSum.max
+        val ymid = scoreSum.indexOf(max)
 
-        case _ =>
+        val (y1, y2) = y.splitAt(ymid)
+        build(x1, y1, builder)
+        build(x2, y2, builder)
       }
     }
+    val builder = Array.newBuilder[Patch]
+    build(a, b, builder)
+    builder.result()
+  }
 
-    lines.mkString
+  private def nwScore(x: Array[String], y: Array[String]): Array[Int] = {
+    val score = Array.fill(x.length + 1, y.length + 1)(0)
+    for (j <- 1 to y.length)
+      score(0)(j) = score(0)(j-1) + ins(y(j-1))
+    for (i <- 1 to x.length) {
+      score(i)(0) = score(i - 1)(0) + del(x(i-1))
+      for (j <- 1 to y.length) {
+        val scoreSub = score(i - 1)(j - 1) + sub(x(i-1), y(j-1))
+        val scoreDel = score(i - 1)(j) + del(x(i-1))
+        val scoreIns = score(i)(j - 1) + ins(y(j-1))
+        score(i)(j) = scoreSub max scoreDel max scoreIns
+      }
+    }
+    Array.tabulate(y.length + 1)(j => score(x.length)(j))
   }
+
+  private def needlemanWunsch(x: Array[String], y: Array[String], builder: mutable.ArrayBuilder[Patch]): Unit = {
+    val d = 1
+    val F = Array.fill(x.length + 1, y.length + 1)(0)
+    for (i <- x.indices)
+      F(i)(0) = d * i
+    for (j <- y.indices)
+      F(0)(j) = d * j
+    for (i <- 1 to x.length) {
+      for (j <- 1 to y.length) {
+        val mtch = F(i - 1)(j - 1) + similarity(x(i-1), y(j-1))
+        val delete = F(i - 1)(j) + d
+        val insert = F(i)(j - 1) + d
+        F(i)(j) = mtch max insert max delete
+      }
+    }
```

Comment:
It was to remain as close as possible to the pseudocode of the algorithm. I will change it now to save up two small traversals.